File: 030/cpp/main.cpp
class Solution {
public:
    vector<int> slide(string& s, map<string, int>& cnt0, int wordLen, int totalLen, int start) {
        map<string, int> cnt;
        vector<int> res;
        for (int i = start, j = start; j < s.size() - wordLen + 1; j += wordLen) {
            string word = s.substr(j, wordLen);
            if (cnt0.find(word) == cnt0.end()) {
                cnt.clear();
                i = j + wordLen;
                continue;
            }
            for (; cnt[word] == cnt0[word] && i < j; i += wordLen) {
                cnt[s.substr(i, wordLen)] -= 1;
            }
            cnt[word] += 1;
            if (j - i + wordLen == totalLen) {
                res.push_back(i);
            }
        }
        return res;
    }

    vector<int> findSubstring(string s, vector<string>& words) {
        int wordLen = words[0].size();
        int totalLen = words.size() * wordLen;
        map<string, int> cnt;
        for (auto &word: words) {
            cnt[word] += 1;
        }
        
        vector<int> ress;
        for (int i = 0; i < wordLen; i++) {
            vector<int> res = slide(s, cnt, wordLen, totalLen, i);
            for (auto &r: res) {
                ress.push_back(r);
            }
        }
        return ress;
    }
};

File: 030/rust/main.rs
use std::collections::HashMap;

impl Solution {
    fn slide(s: &str, cnt0: &mut HashMap<String, usize>, wordLen: usize, totalLen: usize, start: usize) -> Vec<usize> {
        let mut cnt = HashMap::new();
        let (mut i, mut j) = (start, start);
        let mut res = Vec::new();
        while j < s.len() - wordLen + 1 {
            let word = &s[j..j + wordLen];
            let t0 = cnt0.entry(word.to_string()).or_insert(0);
            if (*t0 == 0) {
                cnt.clear();
                i = j + wordLen;
                j += wordLen;
                continue;
            }
            while *cnt.entry(word).or_insert(0) == *t0 {
                let word1 = &s[i..i + wordLen];
                *cnt.entry(word1).or_insert(0) -= 1;
                i += wordLen;
            }
            *cnt.entry(word).or_insert(0) += 1;
            if j - i + wordLen == totalLen {
                res.push(i);
            }
            j += wordLen;
        }
        return res;
    }

    pub fn find_substring(s: String, words: Vec<String>) -> Vec<i32> {
        let wordLen = words[0].len();
        let totalLen = words.len() * wordLen;
        let mut cnt0 = HashMap::new();
        for word in words {
            *cnt0.entry(word).or_insert(0) += 1;
        }

        let mut ress = Vec::new();
        for start in 0..wordLen {
            let res = Solution::slide(&s, &mut cnt0, wordLen, totalLen, start);
            for r in res {
                ress.push(r as i32);
            }
        }
        return ress;
    }
}
